sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -651,6 +651,23 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+  test("Allow changing outer join to broadcast join even if too many empty partitions" +
+    " on build plan") {
+    withSQLConf(
+      SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+      SQLConf.NON_EMPTY_PARTITION_RATIO_FOR_BROADCAST_JOIN.key -> "0.5") {
+      // `testData` is small enough to be broadcast but has empty partition ratio over the config.
+      withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "80") {
+        val (plan, adaptivePlan) = runAdaptiveAndVerifyResult(
+          "SELECT * FROM (select * from testData where value = '1') td" +
+            " right outer join testData2 ON key = a")
```

Comment:
[optional] how about adding all outerJoinTypes here and making it more rigid
can be done with minimal changes 
```
        Seq("right outer", "left outer", "full outer").foreach { joinType =>
          val (plan, adaptivePlan) = runAdaptiveAndVerifyResult(
            "SELECT * FROM (select * from testData where value = '1') td" +
              s" $joinType join testData2 ON key = a")
          val smj = findTopLevelSortMergeJoin(plan)
          assert(smj.size == 1)
          val bhj = findTopLevelBroadcastHashJoin(adaptivePlan)
          assert(bhj.size == 1)
        }
```

Reply:
Thanks for the feedback but I changing `right outer` here to a `left outer` won't have the same semantics w/o changing the sql.  You can only broadcast the inner side.
I don't believe `full outer` supports BHJ at all